File: env/Parameter.cpp
#include "Parameter.h"

using namespace std;

Parameter::Parameter(std::string name, DataType *type, Parameter::Mode mode) {
    this->name = name;
    this->type = type;
    this->mode = mode;
}

Parameter::~Parameter() {
    delete this->type;
}

string Parameter::get_name() {
    return this->name;
}

DataType *Parameter::get_type() {
    return this->type;
}

Parameter::Mode Parameter::get_mode() {
    return this->mode;
}

string Parameter::signature() {
    return this->get_type()->signature(this->get_name());
}

File: pymilldb/context/Parameter.py
import logging

from .DataType import BaseType

logger = logging.getLogger('Parameter').print_pos


class Parameter(object):
    INPUT_MODE = 'IN'
    OUTPUT_MODE = 'OUT'

    mode = None

    @property
    def is_input(self):
        return self.mode == self.INPUT_MODE

    @property
    def is_output(self):
        return self.mode == self.OUTPUT_MODE

    def __init__(self, name: str, kind: BaseType):
        self.name = name
        self.kind = kind

    @property
    def signature(self):
        return self.kind.signature(self.name)


class InputParameter(Parameter):
    mode = Parameter.INPUT_MODE


class OutputParameter(Parameter):
    mode = Parameter.OUTPUT_MODE
